ti/savings/cache/gradle/parser.go

```diff
@@ -0,0 +1,249 @@
+package gradle
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/harness/ti-client/types"
+	"github.com/mattn/go-zglob"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/html"
+)
+
+func ParseSavings(workspace string, log *logrus.Logger, logErr bool) (types.IntelligenceExecutionState, int, error) {
+	cacheState := types.FULL_RUN
+	totalBuildTime := 0
+
+	path := fmt.Sprintf("%s/build/reports/profile/*.html", workspace)
+	files, err := zglob.Glob(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return cacheState, totalBuildTime, fmt.Errorf("no profiles present")
+		}
+		if logErr {
+			log.WithError(err).WithField("path", path).
+				Errorln("errored while trying to resolve path regex")
+		}
+		return cacheState, totalBuildTime, err
+	}
+	if len(files) == 0 {
+		return cacheState, totalBuildTime, fmt.Errorf("no profiles present")
+	}
+
+	foundProfile := false
+	for _, file := range files {
+		htmlContent, err := readHTMLFromFile(file)
+		if err != nil {
+			return cacheState, totalBuildTime, err
+		}
+		reader := strings.NewReader(htmlContent)
+		doc, err := html.Parse(reader)
+		if err != nil {
+			return cacheState, totalBuildTime, err
+		}
+		buildTime, err := parseBuildTimeFromProfile(doc)
+		if err == nil {
+			foundProfile = true
+			totalBuildTime += buildTime
+			if strings.Contains(htmlContent, "FROM-CACHE") {
+				cacheState = types.OPTIMIZED
+			}
+		}
+	}
+	if !foundProfile {
+		return cacheState, totalBuildTime, errors.New("no valid gradle profile html found")
+	}
+	return cacheState, totalBuildTime, nil
+}
+
+func parseBuildTimeFromProfile(n *html.Node) (int, error) {
+	bodyElement := getElementFromNode(n, "body")
+	if bodyElement == nil {
+		return 0, errors.New("no body present")
+	}
+	contentDiv := getDivByID(bodyElement, "content")
+	if contentDiv == nil {
+		return 0, errors.New("no content present")
+	}
+	tabs := getDivByID(contentDiv, "tabs")
+	if tabs == nil {
+		return 0, errors.New("no tabs present")
+	}
+
+	// Summary Div
+	summaryDiv := getDivByID(tabs, "tab0")
+	if summaryDiv == nil {
+		return 0, errors.New("no summary present")
+	}
+	buildTimeStr := extractValueFromTable(summaryDiv, "Total Build Time")
+	totalBuildTime := parseGradleVerseTimeMs(buildTimeStr)
+
+	return totalBuildTime, nil
+}
+
+func extractValueFromTable(n *html.Node, description string) (value string) {
+	if n == nil {
+		return value
+	}
+	var traverse func(*html.Node)
+	traverse = func(node *html.Node) {
```

Comment:
Can this function be written better and with some comments?

Reply:
Updated with comment. Lmk what could be written better